### PR TITLE
feat: Add a metrics for hbase<>bigtable timestamp overflow

### PR DESCRIPTION
**Pull request by sushanb**
Based on PR https://github.com/googleapis/java-bigtable-hbase/pull/3522/files

Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/java-bigtable-hbase/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> ☕️

### CODE REVIEW

File: hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-replication-core/src/main/java/com/google/cloud/bigtable/hbase/replication/adapters/IncompatibleMutationAdapter.java [hunk 1]
```diff
@@ -99,6 +106,11 @@ private boolean isValidDelete(Cell delete) {
     List<Cell> returnedCells = new ArrayList<>(cellsToAdapt.size());
     for (int index = 0; index < cellsToAdapt.size(); index++) {
       Cell cell = cellsToAdapt.get(index);
+      // check whether there is timestamp overflow from HBase -> CBT
+      if (cell.getTimestamp() >= HBASE_EFFECTIVE_MAX_TIMESTAMP) {
```
[the remark this answers is not in the document]
Comment (sushanb) [reply hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-replication-core/src/main/java/com/google/cloud/bigtable/hbase/replication/adapters/IncompatibleMutationAdapter.java]: Done.

File: hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-replication-core/src/main/java/com/google/cloud/bigtable/hbase/replication/adapters/IncompatibleMutationAdapter.java [hunk 2]
```diff
@@ -47,6 +50,8 @@
   private final Connection connection;
   private final Configuration conf;
   private final MetricsExporter metricsExporter;
+  // Maximum timestamp that hbase can send to bigtable in ms.
+  static final long HBASE_EFFECTIVE_MAX_TIMESTAMP = Long.MAX_VALUE / 1000L;
```
Comment (vermas2012) [inline hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-replication-core/src/main/java/com/google/cloud/bigtable/hbase/replication/adapters/IncompatibleMutationAdapter.java:54]: nit: Rename it to Bigtable_Effective_max timestamp. This is a bigtable limitation and has nothing to do wiht HBase.

File: hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-replication-core/src/main/java/com/google/cloud/bigtable/hbase/replication/adapters/IncompatibleMutationAdapter.java [hunk 3]
```diff
@@ -99,6 +114,21 @@ private boolean isValidDelete(Cell delete) {
     List<Cell> returnedCells = new ArrayList<>(cellsToAdapt.size());
     for (int index = 0; index < cellsToAdapt.size(); index++) {
       Cell cell = cellsToAdapt.get(index);
+      // check whether there is timestamp overflow from HBase -> CBT and make sure
+      // it does clash with valid delete which require the timestmap to be
+      // HConstants.LATEST_TIMESTAMP,
+      if (cell.getTimestamp() >= HBASE_EFFECTIVE_MAX_TIMESTAMP
+          && cell.getTimestamp() != HConstants.LATEST_TIMESTAMP) {
+        incrementTimestampOverflowMutations();
+        LOG.error(
```
Comment (vermas2012) [inline hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-replication-core/src/main/java/com/google/cloud/bigtable/hbase/replication/adapters/IncompatibleMutationAdapter.java:123]: We should not log here, if customer is doing reverse timestamps, this can be all the PUTs. If you want to log it. log it in trace mode. 

So you can log this and future puts in trace, users can enable it in test if they want to. Be very prescriptive that DO NOT enable trace logging in prod.

File: hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-replication-core/src/main/java/com/google/cloud/bigtable/hbase/replication/metrics/HBaseToCloudBigtableReplicationMetrics.java [hunk 4]
```diff
@@ -24,4 +24,9 @@ private HBaseToCloudBigtableReplicationMetrics() {}
   public static final String INCOMPATIBLE_MUTATION_METRIC_KEY = "bigtableIncompatibleMutations";
   public static final String DROPPED_INCOMPATIBLE_MUTATION_METRIC_KEY =
       "bigtableDroppedIncompatibleMutations";
+
+  public static final String INCOMPATIBLE_MUTATION_DELETES_METRICS_KEY =
+      "bigtableIncompatibleDeleteMutations";
+  public static final String INCOMPATIBLE_MUTATION_TIMESTAMP_OVERFLOW_METRIC_KEY =
+      "bigtableIncompatibleMutationsTimestampOverflow";
```
Comment (vermas2012) [inline hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-replication-core/src/main/java/com/google/cloud/bigtable/hbase/replication/metrics/HBaseToCloudBigtableReplicationMetrics.java:31]: Please be consistent in naming. It should either be bigtableIncompatibleTimestampOverflowMutation or other one should be bigtableIncompatibleMutationDelete

Comment (vermas2012) [reply hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-replication-core/src/main/java/com/google/cloud/bigtable/hbase/replication/metrics/HBaseToCloudBigtableReplicationMetrics.java:31]: I prefer bigtableIncompatibleTimestampOverflowMutation

File: hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-replication-core/src/test/java/com/google/cloud/bigtable/hbase/replication/adapters/IncompatibleMutationAdapterTest.java [hunk 5]
```diff
@@ -187,7 +194,10 @@ public void testDeleteCanCreateManyDeletes() {
 
     verify(metricsExporter).incCounters(INCOMPATIBLE_MUTATION_METRIC_KEY, 0);
     verify(metricsExporter, times(1)).incCounters(INCOMPATIBLE_MUTATION_METRIC_KEY, 1);
+    verify(metricsExporter, times(1)).incCounters(INCOMPATIBLE_MUTATION_DELETES_METRICS_KEY, 1);
     verify(metricsExporter, times(1)).incCounters(DROPPED_INCOMPATIBLE_MUTATION_METRIC_KEY, 0);
+    verify(metricsExporter, times(0))
```
Comment (vermas2012) [inline hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-replication-core/src/test/java/com/google/cloud/bigtable/hbase/replication/adapters/IncompatibleMutationAdapterTest.java:199]: You can just call NoMoreInteractions on metrics exporter rather than calling expect(0) for all metrics.

Comment (vermas2012) [reply hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-replication-core/src/test/java/com/google/cloud/bigtable/hbase/replication/adapters/IncompatibleMutationAdapterTest.java:199]: https://www.logicbig.com/tutorials/unit-testing/mockito/verify-no-more-interactions.html.

File: hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-replication-core/src/test/java/com/google/cloud/bigtable/hbase/replication/adapters/IncompatibleMutationAdapterTest.java [hunk 6]
```diff
@@ -222,8 +234,34 @@ public void testIncompatibleDeletesAreDropped() {
...
+    walEntryCells.add(put1);
+    walEntryCells.add(put2);
+    walEntryCells.add(delete);
+    BigtableWALEntry walEntry = new BigtableWALEntry(100L, walEntryCells, tableName);
+
+    Assert.assertEquals(
+        Arrays.asList(put1, put2, delete),
+        incompatibleMutationAdapter.adaptIncompatibleMutations(walEntry));
+
+    verify(metricsExporter, times(0)).incCounters(DROPPED_INCOMPATIBLE_MUTATION_METRIC_KEY, 1);
```
Comment (vermas2012) [inline hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-replication-core/src/test/java/com/google/cloud/bigtable/hbase/replication/adapters/IncompatibleMutationAdapterTest.java:261]: please use nomoreinteractions.

**Review by vermas2012 — CHANGES_REQUESTED** [review]

File: hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-replication-core/src/test/java/com/google/cloud/bigtable/hbase/replication/adapters/IncompatibleMutationAdapterTest.java [hunk 7]
```diff
@@ -225,9 +233,42 @@ public void testIncompatibleDeletesAreDropped() {
...
+        incompatibleMutationAdapter.adaptIncompatibleMutations(walEntry));
+
+    testInitialsMetrics();
+    verify(metricsExporter, times(1))
+        .incCounters(INCOMPATIBLE_MUTATION_TIMESTAMP_OVERFLOW_METRIC_KEY, 1);
+    verify(metricsExporter, times(1)).incCounters(INCOMPATIBLE_MUTATION_METRIC_KEY, 1);
+    verifyNoMoreInteractions(metricsExporter);
+  }
+
+  private void testInitialsMetrics() {
```
Comment (vermas2012) [inline hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-replication-core/src/test/java/com/google/cloud/bigtable/hbase/replication/adapters/IncompatibleMutationAdapterTest.java:267]: I don't think we need this method. We can make the tests more DAMP [go/tott/598](https://goto.google.com/tott/598)

**Review by vermas2012 — CHANGES_REQUESTED** [review]
Please add a test with overflow timestamp on 2.x and 1.x integ tests. You can update an existing test or add a new test, up to you.

File: hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-replication-core/src/main/java/com/google/cloud/bigtable/hbase/replication/adapters/IncompatibleMutationAdapter.java [hunk 8]
```diff
@@ -104,6 +121,22 @@ private boolean isValidDelete(Cell delete) {
     List<Cell> returnedCells = new ArrayList<>(cellsToAdapt.size());
     for (int index = 0; index < cellsToAdapt.size(); index++) {
       Cell cell = cellsToAdapt.get(index);
+      // check whether there is timestamp overflow from HBase -> CBT and make sure
+      // it does clash with valid delete which require the timestamp to be
+      // HConstants.LATEST_TIMESTAMP,
+      // this will be true for reverse timestamps. do not enable trace logging
```
[the remark this answers is not in the document]
Comment (sushanb) [reply hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-replication-core/src/main/java/com/google/cloud/bigtable/hbase/replication/adapters/IncompatibleMutationAdapter.java]: Done.

File: hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-replication-core/src/test/java/com/google/cloud/bigtable/hbase/replication/adapters/IncompatibleMutationAdapterTest.java (hunk 7 above)
Comment (sushanb) [reply hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-replication-core/src/test/java/com/google/cloud/bigtable/hbase/replication/adapters/IncompatibleMutationAdapterTest.java]: Done.

File: hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-1.x-replication/src/test/java/com/google/cloud/bigtable/hbase1_x/replication/HbaseToCloudBigtableReplicationEndpointTest.java [hunk 9]
```diff
@@ -414,4 +414,29 @@ public void testWriteFailureToBigtableDoesNotStallReplication()
...
+          return TestReplicationEndpoint.replicatedEntries.get() >= 1;
+        });
+
+    List<Cell> expectedCells =
+        hbaseTable.get(new Get(TestUtils.ROW_KEY).setMaxVersions()).listCells();
+    List<Cell> actualCells = cbtTable.get(new Get(TestUtils.ROW_KEY).setMaxVersions()).listCells();
+
+    for (int i = 0; i < expectedCells.size(); i++) {
+      Assert.assertNotEquals(
+          "Timestamp mismatch for row " + TestUtils.ROW_KEY,
```
Comment (vermas2012) [inline hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-1.x-replication/src/test/java/com/google/cloud/bigtable/hbase1_x/replication/HbaseToCloudBigtableReplicationEndpointTest.java:437]: Timestamps should not match?

File: hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-1.x-replication/src/test/java/com/google/cloud/bigtable/hbase1_x/replication/HbaseToCloudBigtableReplicationEndpointTest.java [hunk 10]
```diff
@@ -414,4 +414,29 @@ public void testWriteFailureToBigtableDoesNotStallReplication()
...
+        () -> {
+          // 1put
+          return TestReplicationEndpoint.replicatedEntries.get() >= 1;
+        });
+
+    List<Cell> expectedCells =
+        hbaseTable.get(new Get(TestUtils.ROW_KEY).setMaxVersions()).listCells();
+    List<Cell> actualCells = cbtTable.get(new Get(TestUtils.ROW_KEY).setMaxVersions()).listCells();
+
+    for (int i = 0; i < expectedCells.size(); i++) {
```
Comment (vermas2012) [inline hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-1.x-replication/src/test/java/com/google/cloud/bigtable/hbase1_x/replication/HbaseToCloudBigtableReplicationEndpointTest.java:435]: Check size =1 and compare result[0]

File: hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-2.x-replication/src/test/java/com/google/cloud/bigtable/hbase2_x/replication/HbaseToCloudBigtableReplicationEndpointTest.java [hunk 11]
```diff
@@ -425,4 +425,29 @@ public void testWriteFailureToBigtableDoesNotStallReplication()
...
+          // 1put
+          return TestReplicationEndpoint.replicatedEntries.get() >= 1;
+        });
+
+    List<Cell> expectedCells =
+        hbaseTable.get(new Get(TestUtils.ROW_KEY).setMaxVersions()).listCells();
+    List<Cell> actualCells = cbtTable.get(new Get(TestUtils.ROW_KEY).setMaxVersions()).listCells();
+
+    for (int i = 0; i < expectedCells.size(); i++) {
+      Assert.assertNotEquals(
```
Comment (vermas2012) [inline hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-2.x-replication/src/test/java/com/google/cloud/bigtable/hbase2_x/replication/HbaseToCloudBigtableReplicationEndpointTest.java:447]: same, don't use array.

**Review by vermas2012 — APPROVED** [review]

File: hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-1.x-replication/src/test/java/com/google/cloud/bigtable/hbase1_x/replication/HbaseToCloudBigtableReplicationEndpointTest.java (hunk 10 above)
Comment (sushanb) [reply hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-1.x-replication/src/test/java/com/google/cloud/bigtable/hbase1_x/replication/HbaseToCloudBigtableReplicationEndpointTest.java]: Done.

File: hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-2.x-replication/src/test/java/com/google/cloud/bigtable/hbase2_x/replication/HbaseToCloudBigtableReplicationEndpointTest.java (hunk 11 above)
Comment (sushanb) [reply hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-2.x-replication/src/test/java/com/google/cloud/bigtable/hbase2_x/replication/HbaseToCloudBigtableReplicationEndpointTest.java]: Done.
